Device runtime support for uart ns16550 and power domain support for siwx91x soc

smalae opened the pull request:
1. Added power domain node in siwg917.dtsi.
2. Updated UART device nodes to reference the newly added power domain.
3. Implemented the pm_action function to manage power domain transitions for the SoC.
4. Enable the pm device runtime driver support for the uart_ns16550 driver (only for devices that have an associated power domain enabled).

smalae committed:
drivers: power_domain: siwx91x: Add power domain driver for siwx91x SoC

1. Added siwx91x power domain node in siwg917.dtsi
2. Updated UART device nodes to reference the newly added power domain.
3. Implemented power domain driver to manage power domain transitions
   for the SoC.

Signed-off-by: Sai Santhosh Malae <Santhosh.Malae@silabs.com>

smalae committed:
drivers: serial: uart_ns16550: Add pm support for uart_ns16550 driver

This commit enables the pm device runtime driver support
for the uart_ns16550 driver (only for devices that have an
associated power domain enabled).

Signed-off-by: Sai Santhosh Malae <Santhosh.Malae@silabs.com>

bjarki-andreasen reviewed (changes requested):
Nice progress!

smalae commented on soc/silabs/silabs_siwx91x/siwg917/soc_siwx91x_power_domain.c:
```diff
@@ -0,0 +1,36 @@
+	case PM_DEVICE_ACTION_TURN_OFF:
+		break;
+	default:
+		ret = -ENOTSUP;
```
Sure

asmellby commented on dts/arm/silabs/siwg917.dtsi:
```diff
@@ -125,13 +125,22 @@
 			status = "disabled";
 		};
 
+		siwx91x_soc_pd: siwx91x_soc_pd {
```
Node labels should have underscore, but node names should use dashes, according to https://docs.zephyrproject.org/latest/contribute/style/devicetree.html

So this should be `siwx91x_soc_pd: siwx91x-soc-pd {`.

smalae replied:
Updated

asmellby commented on dts/arm/silabs/siwg917.dtsi:
```diff
@@ -142,6 +151,7 @@
 			reg-shift = <2>;
 			clocks = <&clock0 SIWX91X_CLK_UART0>;
 			current-speed = <115200>;
+			power-domains = <&siwx91x_soc_pd>;
```
Shouldn't all the UART nodes get `zephyr,pm-device-runtime-auto`?

smalae replied:
Yes, added to the remaining UART nodes

smalae committed:
modules: hal_silabs: Remove CMSIS references

Removed CMSIS references from wiseconnect CMakeLists
as they are no longer required with UART device runtime
PM. Keeping them adds unnecessary compilation overhead.

Signed-off-by: Sai Santhosh Malae <Santhosh.Malae@silabs.com>

smalae committed:
west.yml: Update hal_silabs version

Update hal_silabs version to latest commit:
edbd418670ad8c7cad629a09df2ceee0bc23cf6d

Signed-off-by: Sai Santhosh Malae <Santhosh.Malae@silabs.com>

smalae commented on drivers/serial/uart_ns16550.c:
```diff
@@ -787,6 +788,33 @@ static inline void async_timer_start(struct k_work_delayable *work, size_t timeo
+	case PM_DEVICE_ACTION_TURN_ON:
+		return uart_ns16550_configure(dev, uart_cfg);
+	case PM_DEVICE_ACTION_TURN_OFF:
+		ret = clock_control_off(dev_cfg->clock_dev, dev_cfg->clock_subsys);
```
In that case, we should not perform any operations in 'PM_DEVICE_ACTION_TURN_OFF'. I've already handled that in my latest changes. But let me know if that can be handled differently.

asmellby commented on dts/arm/silabs/siwg917.dtsi:
```diff
@@ -125,13 +125,22 @@
 		};
 
+		siwx91x_soc_pd: siwx91x-soc-pd {
+			compatible = "power-domain";
```
Since the power domain needs a dedicated driver, shouldn't there be a specific binding for SiWx91x power domain?

asmellby commented on soc/silabs/silabs_siwx91x/siwg917/soc_siwx91x_power_domain.c:
```diff
@@ -0,0 +1,34 @@
+#include <zephyr/pm/device.h>
+#include <zephyr/pm/device_runtime.h>
+
+#define SI91X_POWER_DOMAIN DT_NODELABEL(siwx91x_soc_pd)
```
With a dedicated binding, this should set `DT_DRV_COMPAT` instead

asmellby commented on soc/silabs/silabs_siwx91x/siwg917/soc_siwx91x_power_domain.c:
```diff
@@ -0,0 +1,34 @@
+}
+
+PM_DEVICE_DT_DEFINE(SI91X_POWER_DOMAIN, domain_pm_action);
+DEVICE_DT_DEFINE(SI91X_POWER_DOMAIN, NULL, PM_DEVICE_DT_GET(SI91X_POWER_DOMAIN), NULL, NULL,
```
...and this can be `DEVICE_DT_INST_DEFINE()`

asmellby commented on soc/silabs/silabs_siwx91x/siwg917/soc_siwx91x_power_domain.c:
I believe this file should be in `drivers/power_domain/`, not in the `soc` directory

smalae replied:
Added a new driver for siwx91x pd

fabiobaltieri reviewed (approved):
Cool, tested in in a couple platforms using this driver (it82002aw and rts5912) and they still work. Thanks.
